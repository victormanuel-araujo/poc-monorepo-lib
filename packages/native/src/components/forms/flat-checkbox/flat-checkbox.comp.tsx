import React, { useEffect, useState } from 'react'

import { Icon, Text, View } from '@components'
import { TouchableOpacity } from 'react-native'

import {
  CheckboxWrapper,
  UncheckedContainer,
  DisabledCheckboxContainer,
  DisabledCheckboxSignal,
  ChildrenWrapper,
  FlatCheckboxContentWrapper,
} from './flat-checkbox.styles'
import { FlatCheckboxProps } from './flat-checkbox.types'

/**
 * Flat Checkbox component, is a simpler checkbox, pretty close to the Checkbox component but without a card
 *
 * @example
 *
 * ```typescript
 * <FlatCheckbox onChange={() => {}}>
 * 	Text to be rendered
 * </FlatCheckbox>
 * ```
 */
export const FlatCheckbox: React.FC<FlatCheckboxProps> = (props) => {
  /**
   * Destructuring props so it's easier to use each prop when necessary
   */
  const {
    size,
    radius,
    children,
    disabled,
    onChange,
    textProps,
    iconProps,
    textSpacing,
    checkedProps,
    uncheckedProps,
    checked: checkedProp,
    ...accessibilityProps
  } = props

  /**
   * Creating the state based on the `checked` variable passed by props to handle the
   * layout necessary changes
   */
  const [checked, setChecked] = useState<boolean>(checkedProp)

  /**
   * This useEffect is always looking for changes on the checked variable received by props.
   * If any update is made to the checked prop, the internal state should update along with it
   */
  useEffect(() => {
    setChecked(() => checkedProp)
  }, [checkedProp])

  /**
   * The below function handles the onPress method to set if it should check or uncheck, also validate
   * the disabled prop, so when is disabled, nothing is done
   */
  const onCheckboxPress = () => {
    if (disabled) return

    setChecked((prevState) => {
      onChange?.(!prevState)
      return !prevState
    })
  }

  return (
    <TouchableOpacity
      testID="flat-checkbox-comp"
      onPress={onCheckboxPress}
      activeOpacity={disabled ? 1 : 0.8}
      accessibilityRole="checkbox"
      {...accessibilityProps}
    >
      <FlatCheckboxContentWrapper disabled={disabled}>
        {!disabled ? (
          <CheckboxWrapper testID="flat-checkbox-enabled-wrapper" radius={radius} checked={checked} {...checkedProps}>
            {checked ? (
              <Icon.Material
                testID="flat-checkbox-checked-icon"
                icon="check"
                shape="none"
                width={size}
                height={size}
                size={size - 4}
                {...checkedProps}
                {...iconProps}
              />
            ) : (
              <UncheckedContainer
                testID="flat-checkbox-unchecked-comp"
                size={size}
                radius={radius}
                {...uncheckedProps}
              />
            )}
          </CheckboxWrapper>
        ) : (
          <DisabledCheckboxContainer testID="flat-checkbox-disabled-wrapper" size={size} radius={radius}>
            {checked ? (
              <Icon.Material
                testID="flat-checkbox-disabled-checked-icon"
                icon="check"
                shape="none"
                width={size}
                height={size}
                size={size - 4}
                color="gray"
              />
            ) : (
              <DisabledCheckboxSignal testID="flat-checkbox-disabled-unchecked-comp" />
            )}
          </DisabledCheckboxContainer>
        )}

        <ChildrenWrapper textSpacing={textSpacing} disabled={disabled} testID="flat-checkbox-children-wrapper">
          {children.constructor === String ? (
            <Text {...textProps}>{children}</Text>
          ) : (
            <View>{children as JSX.Element}</View>
          )}
        </ChildrenWrapper>
      </FlatCheckboxContentWrapper>
    </TouchableOpacity>
  )
}

FlatCheckbox.defaultProps = {
  radius: 4,
  size: 24,
  checked: false,
  disabled: false,
  textSpacing: 'sm',
  textProps: {},
  iconProps: {},
  checkedProps: {
    color: 'white',
    background: 'black',
  },
  uncheckedProps: {
    thickness: 2,
    color: 'black',
  },
}
