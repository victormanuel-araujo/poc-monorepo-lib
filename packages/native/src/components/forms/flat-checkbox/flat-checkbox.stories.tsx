import React from 'react'

import { FlatCheckbox, FlatCheckboxProps, View, Text, Icon } from '@components'
import { Meta, Story } from '@storybook/react'

export default {
  title: 'Forms/Flat Checkbox',
  component: FlatCheckbox,
} as Meta

const Template: Story<FlatCheckboxProps> = (args) => <FlatCheckbox {...args} />

export const Default = Template.bind({})
Default.args = {
  children: 'Lorem Ipsum',
}

export const Checked = Template.bind({})
Checked.args = {
  ...Default.args,
  checked: true,
}

export const DifferentSize = Template.bind({})
DifferentSize.args = {
  ...Default.args,
  size: 32,
}

export const UncheckedProps = Template.bind({})
UncheckedProps.args = {
  ...Default.args,
  size: 32,
  uncheckedProps: { thickness: 5, color: 'primary' },
}

export const CheckedProps = Template.bind({})
CheckedProps.args = {
  ...Default.args,
  checked: true,
  checkedProps: { background: 'primary', color: 'red' },
}

export const DisabledAndUnchecked = Template.bind({})
DisabledAndUnchecked.args = {
  ...Default.args,
  disabled: true,
}

export const DisabledAndChecked = Template.bind({})
DisabledAndChecked.args = {
  ...Default.args,
  disabled: true,
  checked: true,
}

export const DifferentRadius = Template.bind({})
DifferentRadius.args = {
  ...Default.args,
  radius: 10,
}

export const TextSpacingBetweenCheckbox = Template.bind({})
TextSpacingBetweenCheckbox.args = {
  ...Default.args,
  textSpacing: 'lg',
}

export const WithCustomTextProps = Template.bind({})
WithCustomTextProps.args = {
  ...Default.args,
  textProps: { size: 'lg', weight: 'bold' },
}

export const WithCustomIconProps = Template.bind({})
WithCustomIconProps.args = {
  ...Default.args,
  iconProps: { border: 'primary' },
  checked: true,
}

export const WithChildrenAsComponents = Template.bind({})
WithChildrenAsComponents.args = {
  children: (
    <View dir="row" alignment="center">
      <Icon.Material icon="star" shape="none" />
      <Text>Hacker (Desenvolvimento)</Text>
    </View>
  ),
}
