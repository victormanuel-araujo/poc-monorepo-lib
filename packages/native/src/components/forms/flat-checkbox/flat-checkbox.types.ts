import { ThemeColorsName, ThemeColorsShade, ThemePaddings } from '@theme'
import { AccessibilityProps } from 'react-native'

/**
 * Had to specify the path to the types file because if I didn't, the client will not understand the actual type to be used,
 * the exported interface name is the same as the exported by `react-native` and importing from an index file makes typescript confused.
 * I'm not sure if this behavior occurs when it gets published, would be nice to review this later
 */
import { MaterialIconProps } from '../../icons/material/material.types'
import { TextProps } from '../../typography/text/text.types'

interface FlatCheckboxVariantCommonProps {
  /**
   * Specify the color to be used by the variant
   */
  color: ThemeColorsName

  /**
   * Specify the shade to be used by the variant
   *
   * @default "base"
   */
  shade?: keyof ThemeColorsShade
}

/**
 * This interface extends the `color` and `shade` props and apply to the border when the checked state is false.
 */
export interface FlatCheckboxUncheckedBorder extends FlatCheckboxVariantCommonProps {
  /**
   * Determine the border width value for when is unchecked
   *
   * @default 2
   */
  thickness?: number
}

/**
 * This interface extends the `color` and `shade` props and apply to the checked icon when the checked state is true
 */
export interface FlatCheckboxCheckedProps extends FlatCheckboxVariantCommonProps {
  /**
   * Determine background color to render with the checked icon
   *
   * @default "black"
   */
  background: ThemeColorsName

  /**
   * Determine the background shade for the background color to render with the checked icon
   *
   * @default "base"
   */
  backgroundShade?: keyof ThemeColorsShade
}

export interface FlatCheckboxProps extends AccessibilityProps {
  /**
   * The `children` prop it can be a string that will be rendered along with the checkbox, it also can be
   * a single or more components and keeps the same rules as the string
   *
   * @default `undefined`
   */
  children: string | React.ReactNode

  /**
   * The `onChange` prop is a function that needs to receive a boolean value and should return nothing
   * It is required to the full FlatCheckbox component functionality
   *
   * @default `undefined`
   */
  onChange(value: boolean): void

  /**
   * The `checked` prop is a boolean value that determines the first state inside the component
   * and rules every change, so if this prop changes it's value, the component will change too to match it.
   *
   * @default false
   */
  checked?: boolean

  /**
   * The `disable` prop specifies if the component will allow any interaction with it. It changes the layout
   * to show that it is not available to check or uncheck, show the current status and handles the cursor on web
   *
   * @default false
   */
  disabled?: boolean

  /**
   * Specefies how the component should handle when is checked, having the color and shade for the icon and
   * background and backgroundShade for the background itself
   */
  checkedProps?: FlatCheckboxCheckedProps

  /**
   * Specifies how the component should handle when is unchecked, by passing the color and the shade to the
   * border-color style be correctly setted, also has a `thickness` param to set the border-width value.
   */
  uncheckedProps?: FlatCheckboxUncheckedBorder

  /**
   * Change directly the width and height of the checkbox container and set the check icon size too
   *
   * @default 24
   */
  size?: number

  /**
   * Specify the border radius of the checkbox wrapper component
   *
   * @default 4
   */
  radius?: number

  /**
   * Specify the space between the text and the checkbox using the theme padding values
   *
   * @default "sm"
   */
  textSpacing?: keyof ThemePaddings

  /**
   * The `textProps` is an optional prop that accepts all the props the Text component would receive, except for children,
   * and send it to children text wrapper component, everything passed from this props will be applied to the text.
   * In case the children is not a string this prop is not useful.
   *
   * @default {}
   */
  textProps?: Omit<TextProps, 'children'>

  /**
   * The `iconProps` is an optional prop that accepts all the props the Icon.Material component would receive,
   * except for size, width and height, and send it to the check icon component if the disabled prop is false
   *
   * @default {}
   */
  iconProps?: Omit<MaterialIconProps, 'size' | 'width' | 'height' | 'icon'>
}
