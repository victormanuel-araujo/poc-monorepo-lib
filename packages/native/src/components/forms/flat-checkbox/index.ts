export * from './flat-checkbox.comp'
export { FlatCheckboxProps } from './flat-checkbox.types'
