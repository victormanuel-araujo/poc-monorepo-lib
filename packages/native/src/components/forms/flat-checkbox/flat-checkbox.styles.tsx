import { View } from '@components'
import { getThemeColor, ThemedProps } from '@theme'
import { Platform } from 'react-native'
import styled, { css } from 'styled-components/native'

import { FlatCheckboxCheckedProps, FlatCheckboxProps, FlatCheckboxUncheckedBorder } from './flat-checkbox.types'

/**
 * When the component is rendered on web, it should set the cursor correctly
 * @param disabled - boolean
 */
function handleDisabledWeb({ disabled }: ThemedProps<{ disabled?: boolean }>) {
  if (Platform.OS === 'web') {
    return css`
      cursor: ${() => (disabled ? 'not-allowed' : 'pointer')};
    `
  }
}

/**
 * This function handles the correct alignment on mobile devices.
 * When it's rendered on Android or iOS, the text is always a little to the bottom, this should fix is!
 */
function centralizeTextWithCheckbox() {
  const value = Platform.OS === 'web' ? 0 : 2
  return css`
    transform: translateY(-${value}px);
  `
}

// -------------------------------------- //

export const FlatCheckboxContentWrapper = styled(View).attrs(() => ({
  dir: 'row',
  alignment: 'center',
}))`
  ${handleDisabledWeb}
`

export const CheckboxWrapper = styled(View)<FlatCheckboxCheckedProps & Pick<FlatCheckboxProps, 'radius' | 'checked'>>`
  ${({ background, backgroundShade, radius, checked }) => css`
    width: auto;
    border-radius: ${radius}px;
    background-color: ${checked ? getThemeColor(background, backgroundShade) : 'transparent'};
  `}
`

export const UncheckedContainer = styled(View)<
  FlatCheckboxUncheckedBorder & Pick<FlatCheckboxProps, 'size' | 'radius' | 'checked'>
>`
  ${({ size, thickness, color, shade, radius, checked }) => css`
    width: ${size}px;
    height: ${size}px;
    border-radius: ${radius}px;
    border-width: ${checked ? 0 : thickness}px;
    border-color: ${getThemeColor(color, shade)};
  `}
`

export const DisabledCheckboxContainer = styled(View)<Pick<FlatCheckboxProps, 'size' | 'radius'>>`
  ${({ size, radius }) => css`
    width: ${size}px;
    height: ${size}px;
    border-radius: ${radius}px;
    background-color: ${getThemeColor('gray', 'light')};
    border-color: ${getThemeColor('white')};
    justify-content: center;
    align-items: center;
  `}
`
export const DisabledCheckboxSignal = styled(View)`
  border-width: 1px;
  border-color: ${getThemeColor('gray')};
  width: 0px;
  height: 85%;
  transform: rotate(-135deg);
`

export const ChildrenWrapper = styled(View)<Pick<FlatCheckboxProps, 'textSpacing' | 'disabled'>>`
  margin-left: ${({ textSpacing, theme }) => theme.paddings[textSpacing] || 0}px;
  opacity: ${({ disabled }) => (disabled ? 0.7 : 1)};
  ${centralizeTextWithCheckbox}
`
