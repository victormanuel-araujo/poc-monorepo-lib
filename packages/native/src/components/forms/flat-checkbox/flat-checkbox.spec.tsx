import React from 'react'

import { View, Text } from '@components'
import { fireEvent } from '@testing-library/react-native'
import { createTheme, getThemeColor, renderWithTheme } from '@theme'

import { FlatCheckbox } from './flat-checkbox.comp'

describe('Flat Checkbox component', () => {
  const defaultOnChangeFn = jest.fn()

  it('should render correctly and match snapshot', () => {
    const flatCheckbox = renderWithTheme(<FlatCheckbox onChange={defaultOnChangeFn}>Lorem Ipsum</FlatCheckbox>)

    expect(flatCheckbox.getByText('Lorem Ipsum')).toBeTruthy()
    expect(flatCheckbox.toJSON()).toMatchSnapshot()
  })

  it('should render correctly with custom children', () => {
    const flatCheckbox = renderWithTheme(
      <FlatCheckbox onChange={defaultOnChangeFn}>
        <View spacing="lg">
          <Text>Text inside</Text>
        </View>
      </FlatCheckbox>
    )

    expect(flatCheckbox).toBeTruthy()
  })

  it('should call onChange prop function when not disabled and component is pressed', () => {
    const onChangeFn = jest.fn()
    const flatCheckbox = renderWithTheme(<FlatCheckbox onChange={onChangeFn}>Lorem Ipsum</FlatCheckbox>)

    const touchableFlatCheckbox = flatCheckbox.getByTestId('flat-checkbox-comp')
    fireEvent.press(touchableFlatCheckbox)
    fireEvent.press(touchableFlatCheckbox)
    fireEvent.press(touchableFlatCheckbox)
    expect(onChangeFn).toBeCalledTimes(3)
  })

  it('should not call onChange prop function when disabled and component is pressed', () => {
    const onChangeFn = jest.fn()
    const flatCheckbox = renderWithTheme(
      <FlatCheckbox disabled onChange={onChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    )

    const touchableFlatCheckbox = flatCheckbox.getByTestId('flat-checkbox-comp')
    fireEvent.press(touchableFlatCheckbox)
    fireEvent.press(touchableFlatCheckbox)
    fireEvent.press(touchableFlatCheckbox)
    expect(onChangeFn).toBeCalledTimes(0)
  })

  it('should render the checked icon when checked prop is true', () => {
    const checkedIcon = renderWithTheme(
      <FlatCheckbox checked onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    ).getByTestId('flat-checkbox-checked-icon')

    expect(checkedIcon).toBeTruthy()
  })

  it('should not render the checked icon when checked prop is false', () => {
    const uncheckedComp = renderWithTheme(
      <FlatCheckbox onChange={defaultOnChangeFn}>Lorem Ipsum</FlatCheckbox>
    ).getByTestId('flat-checkbox-unchecked-comp')

    expect(uncheckedComp).toBeTruthy()
  })

  it('should render the disabled without the checked icon when disabled is true and checked prop is false', () => {
    const disabledUncheckedComp = renderWithTheme(
      <FlatCheckbox disabled onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    ).getByTestId('flat-checkbox-disabled-unchecked-comp')

    expect(disabledUncheckedComp).toBeTruthy()
  })

  it('should render the disabled with the checked icon when disabled and checked prop is true', () => {
    const disabledCheckedComp = renderWithTheme(
      <FlatCheckbox disabled checked onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    ).getByTestId('flat-checkbox-disabled-checked-icon')

    expect(disabledCheckedComp).toBeTruthy()
  })

  it('should respect and apply the checked props when checked', () => {
    const theme = createTheme()
    const flatCheckbox = renderWithTheme(
      <FlatCheckbox checked checkedProps={{ color: 'red', background: 'yellow' }} onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    )

    const checkedIcon = flatCheckbox.getByTestId('flat-checkbox-checked-icon')
    const checkedWrapperComp = flatCheckbox.getByTestId('flat-checkbox-enabled-wrapper')

    expect(checkedIcon).toBeTruthy()
    expect(checkedIcon).toHaveStyle({ color: getThemeColor('red')({ theme }) })
    expect(checkedWrapperComp).toHaveStyle({ backgroundColor: getThemeColor('yellow')({ theme }) })
  })

  it('should respect and apply the unchecked props when not checked', () => {
    const flatCheckbox = renderWithTheme(
      <FlatCheckbox uncheckedProps={{ thickness: 4, color: 'primary' }} onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    )
    const uncheckedComp = flatCheckbox.getByTestId('flat-checkbox-unchecked-comp')

    expect(uncheckedComp).toBeTruthy()
    expect(uncheckedComp).toHaveProp('color', 'primary')
    expect(uncheckedComp).toHaveStyle({ borderWidth: 4 })
  })

  it('should apply text props on the children when is a string', () => {
    const theme = createTheme()
    const flatCheckbox = renderWithTheme(
      <FlatCheckbox textProps={{ color: 'primary', size: 'lg' }} onChange={defaultOnChangeFn}>
        Lorem Ipsum
      </FlatCheckbox>
    )

    const textComponent = flatCheckbox.getByText('Lorem Ipsum')
    expect(textComponent).toBeTruthy()
    expect(textComponent).toHaveStyle({ color: getThemeColor('primary')({ theme }), fontSize: theme.fontSizes.lg })
  })
})
