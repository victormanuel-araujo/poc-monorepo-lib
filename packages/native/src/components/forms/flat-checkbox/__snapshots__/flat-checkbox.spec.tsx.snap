// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flat Checkbox component should render correctly and match snapshot 1`] = `
<View
  accessibilityRole="checkbox"
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="flat-checkbox-comp"
>
  <View
    alignment="center"
    backgroundShade="base"
    dir="row"
    disabled={false}
    display="flex"
    position="relative"
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "paddingBottom": 0,
        "paddingLeft": 0,
        "paddingRight": 0,
        "paddingTop": 0,
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <View
      background="black"
      backgroundShade="base"
      checked={false}
      color="white"
      dir="column"
      display="flex"
      position="relative"
      radius={4}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderRadius": 4,
          "display": "flex",
          "flexDirection": "column",
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "position": "relative",
          "width": "auto",
        }
      }
      testID="flat-checkbox-enabled-wrapper"
    >
      <View
        backgroundShade="base"
        color="black"
        dir="column"
        display="flex"
        position="relative"
        radius={4}
        size={24}
        style={
          Object {
            "borderColor": "#000000",
            "borderRadius": 4,
            "borderWidth": 2,
            "display": "flex",
            "flexDirection": "column",
            "height": 24,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 0,
            "position": "relative",
            "width": 24,
          }
        }
        testID="flat-checkbox-unchecked-comp"
        thickness={2}
      />
    </View>
    <View
      backgroundShade="base"
      dir="column"
      disabled={false}
      display="flex"
      position="relative"
      style={
        Object {
          "display": "flex",
          "flexDirection": "column",
          "marginLeft": 8,
          "opacity": 1,
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
          "position": "relative",
          "transform": Array [
            Object {
              "translateY": -2,
            },
          ],
          "width": "100%",
        }
      }
      testID="flat-checkbox-children-wrapper"
      textSpacing="sm"
    >
      <Text
        accessibilityRole="text"
        style={
          Object {
            "color": "#000000",
            "fontFamily": "Raleway",
            "fontSize": 16,
            "fontWeight": "500",
            "lineHeight": 24,
          }
        }
        type="description"
      >
        Lorem Ipsum
      </Text>
    </View>
  </View>
</View>
`;
